src/main/java/com/pusher/client/channel/impl/ChannelImpl.java

```diff
@@ -111,18 +111,23 @@ public void onMessage(final String event, final String message) {
             if (listeners != null) {
                 for (final SubscriptionEventListener listener : listeners) {
                     final String data = extractDataFrom(message);
-
+                    final PusherEvent metadata = new PusherEvent(getMetaData(message));
```

Comment:
I have no idea why this was. This is why in the initial metadata implementation we sent with string, string which stuck around to this release. If we don't make that assumption, how can the user make any assumptions about the 'extra' data that they're getting?

I can't think of any circumstances where this is the case... @WillSewell any opinion here?

Reply:
> Currently you're parsing the message JSON twice: once in `extractDataFrom` and once in `getMetaData`. This could be improved by modifying `extractDataFrom` to use the output of `getMetaData` perhaps.

w.r.t this, good point. again, the different implementations were hangovers from the metadata implementation.

Reply:
In total, that means the library actually parses each message 4 times. Once to determine the event name: https://github.com/pusher/pusher-websocket-java/blob/1894ed21512b63f3b78bbc6fc04040eab46b24c6/src/main/java/com/pusher/client/connection/websocket/WebSocketConnection.java#L259

Once to determine the channel: https://github.com/pusher/pusher-websocket-java/blob/master/src/main/java/com/pusher/client/channel/impl/ChannelManager.java

Then these two times.

Reply:
I agree with @TomKemp that we should not be deserialising a second time, particularly as it is very easy to share the logic of `extractDataFrom` in this case.

I think the cleanest way of doing this is to allow deserialisation directly into a `PusherEvent`. It looks like you can do this by creating a `PusherEventDeserializer implements JsonDeserializer<PusherEvent>`, which would  have the same logic as `getMetaData` currently has. See "Writing a Deserializer" in https://sites.google.com/site/gson/gson-user-guide. This would let us do something like

```
final PusherEvent pusherEvent = GSON.fromJson(message, PusherEvent.class);
```

You could remove the `extractDataFrom` method and fetch the data from the `pusherEvent`.

Reply:
I also think the whole event should be decoded as `Map<Object, Object>` (instead of  `Map<String, String>`) as this gives us maximum flexibility in the future. The helper methods like `getUserId` can specialise the type.

Maybe @jameshfisher has some thoughts here?

Reply:
`Map<String, Object>` IMO (keys in JSON are always strings)